core/domain/user_jobs_one_off.py

```diff
@@ -470,3 +471,136 @@ def map(user_settings_model):
     def reduce(key, values):
         """Implements the reduce function for this job."""
         yield (key, len(values))
+
+
+class CleanUpUserSubscribersModelOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """Job that cleans up UserSubscribersModel by removing user id if it is
+    present in subscriber ids.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [user_models.UserSubscribersModel]
+
+    @staticmethod
+    def map(item):
+        if not item.deleted and item.id in item.subscriber_ids:
+            item.subscriber_ids.remove(item.id)
+            item.put()
+            yield ('Removed user from their own subscribers list', item.id)
+
+    @staticmethod
+    def reduce(key, values):
+        values.sort()
+        yield (key, values)
+
+
+class CleanUpCollectionProgressModelOneOffJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """Job that cleans up CollectionProgressModel by:
+    1. Removing exploration ids which are not a part of the collection.
+    2. Creating CompletedActivitiesModel for completed explorations if it is
+    missing.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [user_models.CollectionProgressModel]
+
+    @staticmethod
+    def map(item):
+        if not item.deleted:
+            completed_activities_model = (
+                user_models.CompletedActivitiesModel.get_by_id(item.user_id))
+
+            if completed_activities_model is None:
+                completed_activities_model = (
+                    user_models.CompletedActivitiesModel(id=item.user_id))
+                completed_activities_model.exploration_ids = (
+                    item.completed_explorations)
+                completed_activities_model.put()
+                yield ('Regenerated Missing CompletedActivitiesModel', item.id)
+
+            if completed_activities_model.deleted:
```

Comment:
Do we know for certain that, if the CollectionProgressModel is not deleted but this model is, then the eventual status ought to be that both are not deleted? Could it be that the CollectionProgressModel is the one that ought to have been deleted?

Reply:
If collection progress model is deleted, we do not reach this line and no action will be taken.

Also, if the collection progress model was supposed to be deleted, then the collection progress audit job would have resulted in errors regarding the collection being deleted or missing or the user being deleted or missing. But since there were no such errors, it implies that collection progress model should exist since there is a collection being used by the user and the completed activities model is missing.